src/api/app/models/badge.rb

```diff
@@ -0,0 +1,27 @@
+class Badge
+  RED = '#e05d44'.freeze
+  GREEN = '#4c1'.freeze
+
+  def initialize(type)
+    @type = type
+    @filename = 'badge-unknown.svg'
+  end
+
+  def status!(results)
+    @results = results
+    return if @results.blank?
+
+    @filename = 'badge-failed.svg' if @results.any? { |r| r.code == 'failed' }
+    @filename = 'badge-succeeded.svg' if @results.all? { |r| r.code == 'succeeded' }
+    @filename = 'badge-percent.svg' if @type == 'percent'
+  end
+
+  def xml
+    file = Rails.public_path.join(@filename).read
+    return file unless @filename == 'badge-percent.svg'
```

Comment:
Checking for `@type == 'percent'` instead of `@filename == 'badge-percent.svg'` would allow to extract  `@type == 'percent'` into a private helper method and reusing this code here and in the `filename` instance method I wrote about in my other comment above.